PlayFabApiTest.js

```diff
@@ -0,0 +1,420 @@
+            assert.ok(success, message);
+        if (error != null)
+            assert.ok(false, "PlayFab error message: " + error.errorMessage);
+    },
+
+    InvalidLogin: function (assert) {
+        var invalidRequest = {
+            TitleId: PlayFab.settings.titleId,
+            Email: PlayFabApiTests.titleData.userEmail,
+            Password: PlayFabApiTests.titleData.userPassword + "INVALID",
+        };
+
+        var InvalidLoginCallback = function (result, error) {
+            assert.ok(result == null, "Login should have failed");
+            assert.ok(error != null, "Login should have failed");
+            if (error != null)
+                assert.ok(error.errorMessage.toLowerCase().indexOf("password") > -1, "Expect errorMessage about invalid password: " + error.errorMessage);
+            invalidDone();
+        };
+        var invalidDone = assert.async();
+        PlayFabClientSDK.LoginWithEmailAddress(invalidRequest, PlayFabApiTests.CallbackWrapper("InvalidLoginCallback", InvalidLoginCallback, assert));
+    },
+
+    LoginOrRegister: function (assert) {
+        var loginRequest = {
+            // Currently, you need to look up the correct format for this object in the API-docs:
+            //   https://api.playfab.com/Documentation/Client/method/LoginWithEmailAddress
+            TitleId: PlayFab.settings.titleId,
+            Email: PlayFabApiTests.titleData.userEmail,
+            Password: PlayFabApiTests.titleData.userPassword,
+        };
+        var registerRequest = {
+            // Currently, you need to look up the correct format for this object in the API-docs:
+            //   https://api.playfab.com/Documentation/Client/method/RegisterPlayFabUser
+            TitleId: PlayFab.settings.titleId,
+            Username: PlayFabApiTests.titleData.userName,
+            Email: PlayFabApiTests.titleData.userEmail,
+            Password: PlayFabApiTests.titleData.userPassword,
+        };
+
+        // We don't know at this point how many async calls we'll make
+        var loginDone = null;
+        var registerDone = null;
+
+        var OptionalLoginCallback = function (result, error) {
+            // First login falls back upon registration if login failed
+            if (result == null) {
+                // Register the character and try again
+                registerDone = assert.async();
+                PlayFabClientSDK.RegisterPlayFabUser(registerRequest, PlayFabApiTests.CallbackWrapper("RegisterCallback", RegisterCallback, assert));
+                loginDone();
+            }
+            else {
+                // Confirm the successful login
+                MandatoryLoginCallback(result, error)
+            }
+        };
+        var RegisterCallback = function (result, error) {
+            // Second login MUST succeed
+            PlayFabApiTests.VerifyNullError(result, error, assert, "Testing Registration result");
+
+            // Log in again, this time with the newly registered account
```

Comment:
OIC, yes.  I could do that.  I guess I fell into the pattern of: login, [register, login].  For now, I think we should leave it this way, as it matches all other auto-testing in all sdks.  Maybe I can make a pass later to remove the second login call everywhere, as you're right I don't think it's necessary.